Source/GSOrderedSet.m

```diff
@@ -0,0 +1,399 @@
+#define	GSI_ARRAY_NO_RELEASE	1
+#define	GSI_ARRAY_NO_RETAIN	1
+#define GSI_ARRAY_TYPES       GSUNION_OBJ
+
+#define GSI_ARRAY_RELEASE(A, X)	[(X).obj release]
+#define GSI_ARRAY_RETAIN(A, X)	[(X).obj retain]
+
+#import "GNUstepBase/GSIArray.h"
+
+//  static SEL     memberSel;
+static SEL      privateCountOfSel;
+@interface GSOrderedSet : NSOrderedSet
+{
+@public
+  GSIArray_t array;
+}
+@end
+
+@interface GSMutableOrderedSet : NSMutableOrderedSet
+{
+@public
+  GSIArray_t array;
+@private
+  NSUInteger _version;
+}
+@end
+
+@interface GSOrderedSetEnumerator : NSEnumerator
+{
+  GSOrderedSet *set;
+  unsigned      current;
+  unsigned      count;
+}
+@end
+
+@interface GSOrderedSetEnumeratorReverse : GSOrderedSetEnumerator
+@end
+
+@implementation GSOrderedSetEnumerator
+- (id) initWithOrderedSet: (NSOrderedSet*)d
+{
+  self = [super init];
+  if (self != nil)
+    {
+      set = (GSOrderedSet*)RETAIN(d);
+      current = 0;
+      count = GSIArrayCount(&set->array);
+    }
+  return self;
+}
+
+- (id) nextObject
+{
+  if(current < count)
+    {
+      GSIArrayItem item = GSIArrayItemAtIndex(&set->array, current);
+      current++;
+      return (id)(item.obj);
+    }
+  return nil;
+}
+
+- (void) dealloc
+{
+  RELEASE(set);
+  [super dealloc];
+}
+@end
+
+
+@implementation GSOrderedSetEnumeratorReverse
+- (id) initWithOrderedSet: (GSOrderedSet*)d
+{
+  self = [super initWithOrderedSet: d];
+  if(self != nil)
+    {
+      current = GSIArrayCount(&set->array);
+    }
+  return self;
+}
+
+- (id) nextObject
+{
+  GSIArrayItem item;
+
+  if (current == 0)
+    return nil;
+  
+  item = GSIArrayItemAtIndex(&set->array, --current);
+  return (id)(item.obj);
+}
+@end
+
+@implementation GSOrderedSet
+
+static Class	setClass;
+static Class	mutableSetClass;
+
++ (void) initialize
+{
+  if (self == [GSOrderedSet class])
+    {
+      setClass = [GSOrderedSet class];
+      mutableSetClass = [GSMutableOrderedSet class];
+      privateCountOfSel = @selector(_countForObject:);
+    }
+}
+
+- (id) copyWithZone: (NSZone*)z
+{
+  return RETAIN(self);
+}
+
+- (void) dealloc
+{
+  GSIArrayEmpty(&array);
+  [super dealloc];
+}
+
+- (NSUInteger) hash
+{
+  return [self count];
+}
+
+- (id) init
+{
+  return [self initWithObjects: NULL count: 0];
+}
+
+- (NSEnumerator*) objectEnumerator
+{
+  return AUTORELEASE([[GSOrderedSetEnumerator alloc] initWithOrderedSet: self]);
+}
+
+- (NSEnumerator*) reverseObjectEnumerator
+{
+  return AUTORELEASE([[GSOrderedSetEnumeratorReverse alloc] initWithOrderedSet: self]);
+}
+
+- (NSUInteger) countByEnumeratingWithState: (NSFastEnumerationState*)state
+                                   objects: (id*)stackbuf
+                                     count: (NSUInteger)len
+{
+  //state->mutationsPtr = (unsigned long *)self;
+  //return GSIMapCountByEnumeratingWithStateObjectsCount
+  //            (&map, state, stackbuf, len);
+
+  return [self count];
+}
+
+- (NSUInteger) sizeInBytesExcluding: (NSHashTable*)exclude
```

Comment:
Richard just changed the behaviour of this method in the main branch. You may need to have a look at that change.

Reply:
I can look at that.  I'm trying to make the method more automated (so people mostly don't need to override it in subclasses), and while the current code in the main branch works I'm not yet finished with it (I'm wondering whether to use malloc_usable_size() on non-object pointer ivars in the default code).

Reply:
Marking this for later resolution.  I will add an issue and assign rfm.